ext/ddtrace_profiling_native_extension/collectors_idle_sampling_helper.c

```diff
@@ -0,0 +1,241 @@
+  rb_define_singleton_method(collectors_idle_sampling_helper_class, "_native_reset", _native_reset, 1);
+  rb_define_singleton_method(testing_module, "_native_idle_sampling_helper_request_action", _native_idle_sampling_helper_request_action, 1);
+}
+
+// This structure is used to define a Ruby object that stores a pointer to a struct idle_sampling_loop_state
+// See also https://github.com/ruby/ruby/blob/master/doc/extension.rdoc for how this works
+static const rb_data_type_t idle_sampling_helper_typed_data = {
+  .wrap_struct_name = "Datadog::Profiling::Collectors::IdleSamplingHelper",
+  .function = {
+    .dmark = NULL, // We don't store references to Ruby objects so we don't need to mark any of them
+    .dfree = RUBY_DEFAULT_FREE,
+    .dsize = NULL, // We don't track memory usage (although it'd be cool if we did!)
+    //.dcompact = NULL, // Not needed -- we don't store references to Ruby objects
+  },
+  .flags = RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+static VALUE _native_new(VALUE klass) {
+  struct idle_sampling_loop_state *state = ruby_xcalloc(1, sizeof(struct idle_sampling_loop_state));
+
+  reset_state(state);
+
+  return TypedData_Wrap_Struct(klass, &idle_sampling_helper_typed_data, state);
+}
+
+static void reset_state(struct idle_sampling_loop_state *state) {
+  state->wakeup_mutex = (pthread_mutex_t) PTHREAD_MUTEX_INITIALIZER;
+  state->wakeup = (pthread_cond_t) PTHREAD_COND_INITIALIZER;
+  state->requested_action = ACTION_WAIT;
+  state->run_action_function = NULL;
+}
+
+// The same instance of the IdleSamplingHelper can be reused multiple times, and this resets it back to
+// a pristine state before recreating the worker thread (this includes resetting the mutex in case it was left
+// locked halfway through the VM forking)
+static VALUE _native_reset(DDTRACE_UNUSED VALUE self, VALUE self_instance) {
+  struct idle_sampling_loop_state *state;
+  TypedData_Get_Struct(self_instance, struct idle_sampling_loop_state, &idle_sampling_helper_typed_data, state);
+
+  reset_state(state);
+
+  return Qtrue;
+}
+
+static VALUE _native_idle_sampling_loop(DDTRACE_UNUSED VALUE self, VALUE self_instance) {
+  struct idle_sampling_loop_state *state;
+  TypedData_Get_Struct(self_instance, struct idle_sampling_loop_state, &idle_sampling_helper_typed_data, state);
+
+  // Release GVL and run the loop waiting for requests
+  rb_thread_call_without_gvl(run_idle_sampling_loop, state, interrupt_idle_sampling_loop, state);
+
+  return Qtrue;
+}
+
+static void *run_idle_sampling_loop(void *state_ptr) {
+  struct idle_sampling_loop_state *state = (struct idle_sampling_loop_state *) state_ptr;
+  int error = 0;
+
+  while (true) {
+    ENFORCE_SUCCESS_NO_GVL(pthread_mutex_lock(&state->wakeup_mutex));
+
+    action next_action;
+    void (*run_action_function)(void);
+
+    // Await for an action
+    while ((next_action = state->requested_action) == ACTION_WAIT) {
```

Comment:
I'm assuming it's correct to check for an `ACTION_WAIT` immediately on start? (while vs do-while kind of question.)